plutus-tx/src/PlutusTx/SortedMap.hs

```diff
@@ -0,0 +1,78 @@
+{-# LANGUAGE DerivingStrategies #-}
+{-# LANGUAGE ViewPatterns       #-}
+
+{-# OPTIONS_GHC -fno-omit-interface-pragmas #-}
+{-# OPTIONS_GHC -fno-spec-constr #-}
+{-# OPTIONS_GHC -fno-specialise #-}
+
+module PlutusTx.SortedMap
+    ( SortedMap (..)
+    , empty
+    , singleton
+    , unsafeFromListUnique
+    , eqWith
+    ) where
+
+import Prelude qualified as Haskell
+
+import PlutusTx.Base
+import PlutusTx.Ord
+import PlutusTx.Prelude
+
+import Data.Coerce (coerce)
+
+-- | A map from @k@ to @v@ backed by a list that is supposed to be sorted from lowest to highest
+-- @k@s with no key duplicated.
+newtype SortedMap k v = UnsafeSortedMap
+    { unSortedMap :: [(k, v)]
+    } deriving stock (Haskell.Show)
+
+{-# INLINABLE empty #-}
+empty :: SortedMap k v
+empty = UnsafeSortedMap []
+
+{-# INLINE singleton #-}
+singleton :: k -> v -> SortedMap k [v]
+singleton ~k ~v = UnsafeSortedMap [(k, [v])]
+
+{-# INLINE unsafeInsertOneUnique #-}
+-- | Insert a key-value pair into the 'SortedMap' assuming the key isn't already in the map (if it
+-- is, the function throws).
+unsafeInsertOneUnique :: forall k v. Ord k => k -> v -> SortedMap k v -> SortedMap k v
+unsafeInsertOneUnique ~k0 ~v0 = coerce go where
+    go :: [(k, v)] -> [(k, v)]
+    go []                  = [(k0, v0)]
+    go kvs@((k, v) : kvs') =
+        case k0 `compare` k of
+            LT -> (k0, v0) : kvs
+            -- TODO: make this @traceError duplicateElements@.
+            EQ -> (k, v0) : kvs'
```

Comment:
I'm not making the change in this PR, so that the budgets comparison is fair. I'll make the change and I'll update the golden tests in general, since the current ones were written with the assumption that duplicate currencies/tokens are allowed.